### PR TITLE
Change enum->int casts to not go through MIR casts.

### PR DESCRIPTION
**Pull request by oli-obk**
follow-up to https://github.com/rust-lang/rust/pull/96814

this simplifies all backends and even gives LLVM more information about the return value of `Rvalue::Discriminant`, enabling optimizations in more cases.

### CODE REVIEW

**Comment by oli-obk** [comment]
🤣 enum -> int cast acting as mem::forget strikes again. We used to drop these enums right after casting. [Then we got MIR and we ended up `mem::forget`ting them instead.](https://github.com/rust-lang/rust/issues/35941) We do have a lint against them now: https://github.com/rust-lang/rust/issues/73333

File: compiler/rustc_codegen_ssa/src/mir/rvalue.rs
Comment (RalfJung) [inline compiler/rustc_codegen_ssa/src/mir/rvalue.rs:277]: The interpreter contains a similar case that can be removed now.

File: compiler/rustc_mir_build/src/build/expr/as_rvalue.rs
Comment (RalfJung) [inline compiler/rustc_mir_build/src/build/expr/as_rvalue.rs:184]: and what?

File: compiler/rustc_mir_build/src/build/expr/as_rvalue.rs
Comment (RalfJung) [inline compiler/rustc_mir_build/src/build/expr/as_rvalue.rs:185]: `ty.is_enum()` is a thing, that should work here, shouldn't it?

Comment (oli-obk) [reply compiler/rustc_mir_build/src/build/expr/as_rvalue.rs:185]: sure, but we need the `adt_def` inside the block

**Comment by RalfJung** [comment]
> follow-up to https://github.com/rust-lang/rust/pull/96814

To be clear, your PR needs to land first before #96814 can make progress.

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (RalfJung) [inline compiler/rustc_lint_defs/src/builtin.rs:2610]: This looks like it should be a separate PR?

Also, should this be shown in the future compat report as well so people notice when their dependencies still rely on this?

Comment (oli-obk) [reply compiler/rustc_lint_defs/src/builtin.rs:2610]: This does not affect dependencies in a *new* way. We already did report this in future compat reports. All I did change the level in main crates.

Comment (oli-obk) [reply compiler/rustc_lint_defs/src/builtin.rs:2610]: > This looks like it should be a separate PR?

Also this PR flips the logic *again* on what happens to casts from c-like enums with Drop impls. I found it prudent to bump the level :)

Comment (RalfJung) [reply compiler/rustc_lint_defs/src/builtin.rs:2610]: > We already did report this in future compat reports.

I don't think so. This requires a special flag in the `FutureIncompatibleInfo`. See e.g. https://github.com/rust-lang/rust/pull/95372.

Comment (RalfJung) [reply compiler/rustc_lint_defs/src/builtin.rs:2610]: > Also this PR flips the logic again on what happens to casts from c-like enums with Drop impls. I found it prudent to bump the level :)

Sure, I agree. I am just saying that's a decision process which should probably be separate from the rest of this PR.

File: src/test/run-pass-valgrind/cast-enum-with-dtor.rs
Comment (RalfJung) [inline src/test/run-pass-valgrind/cast-enum-with-dtor.rs:33]:
```suggestion
    // When the scope ends, the dtor has been called.
    assert_eq!(FLAG.load(Ordering::SeqCst), 1);
```

**Comment by RalfJung** [comment]
Looks like the perf run didn't go through. Not sure if the old try build is still sufficiently up-to-date, so let's
@bors try @rust-timer queue

**Comment by RalfJung** [comment]
Regarding the PR itself, the change LGTM (not that I feel qualified to review this), but we have two decisions to make:
- Are we okay with regression bounds checks for array accesses derived from enum casts? Not sure how common that pattern is, but someone cared enough to add a test. OTOH this is on the critical path for an ICE fix (https://github.com/rust-lang/rust/issues/92464, https://github.com/rust-lang/rust/issues/96185) and all backend people seem to agree that casting is better done in MIR (or else all backends have to duplicate the same logic, as they did before this PR).
- Are we okay with changing dtor order on C-like enums with `Drop` again? We warn against these casts for quite a while already (https://github.com/rust-lang/rust/issues/73333) and have changed behavior in the past so I hope the answer is yes.

Cc @rust-lang/compiler for the first point; @rust-lang/lang for the second.

File: src/test/codegen/enum-bounds-check-derived-idx.rs
Comment (erikdesjardins) [inline src/test/codegen/enum-bounds-check-derived-idx.rs:19]: I think what we really need here is range attributes on parameters (which LLVM doesn't currently support).

Presumably, this PR generates IR like:
```ll
define @lookup_inc(..., i8 %f) {
  %0 = alloca i8
  store i8 %f, ptr %0
  %1 = load i8, ptr %0, !range !0
  ...
}

!0 = !{i8 0, i8 4}
```

But that will get optimized early on to something like:

```ll
define @lookup_inc(..., i8 %f) {
  %1 = %f
  ...
}
```
...dropping the range metadata on the floor.

One way to avoid this is to add assumes, like we did before. LLVM itself could also, in theory, add assumes in mem2reg to retain knowledge from allocas it drops.

But it seems like the real underlying problem is the lack of range attributes on parameters. If we could just generate IR like:

```ll
define @lookup_inc(..., i8 range(!0) %f) {
  ...
}

!0 = !{i8 0, i8 4}
```

...then we wouldn't need assumes or fragile `!range` metadata on alloca loads in the body of the function.

I will look into this, although I'm slightly worried that the reason `range` attributes aren't supported on parameters is because parameter attributes currently only accept 0 or 1 integer arguments, and it would take extensive refactoring to make them accept multiple arguments or reference a metadata node. Let's hope that's not the case.

**Comment by oli-obk** [comment]
If we want this optimization in a backend-agnostic way, we should probably add a `StatmentKind::Assume` and introduce that in a few places.

@bors try @rust-timer queue

**Comment by oli-obk** [comment]
Hmm... do we have any regex runtime benchmarks? LLVM may be producing faster code now for non-cast cases

Or it just has to process more `assume`s without producing better machine code

Opinions? Ideas?

**Comment by nagisa** [comment]
I would suggest that any changes that focus on potential runtime performance improvements due to assumes are a part of a separate PR.

Sadly you will need to run benchmarks locally to estimate the improvement there.

**Comment by oli-obk** [comment]
> I would suggest that any changes that focus on potential runtime performance improvements due to assumes are a part of a separate PR.

The only way I see to keep the current behaviour without also introducing new `assume`s is to introduce `assume`s at the mir building stage. They are intrinsic function calls right now, so that would be very disruptive. I could do a preliminary PR that adds an `Assume` statement to MIR and lower all intrinsic function calls to it, just like we do with several other intrinsics. Then in this PR I could change the MIR building to introduce said `assume`s. This would also enable the opt for backends other than LLVM.

This could also be done after merging this PR as is (with the perf regression).

> Sadly you will need to run benchmarks locally to estimate the improvement there.

I'll try to do that

**Comment by joshtriplett** [comment]
@RalfJung I think we've settled the lang question in https://github.com/rust-lang/rust/pull/97652 ; please re-nominate if there's an additional question that doesn't cover.

**Comment by oli-obk** [comment]
> Sadly you will need to run benchmarks locally to estimate the improvement there.

I tried this locally. Compared with nightly most benchmarks are a bit faster but there are some massive regressions (e.g. `misc::long_needle1` is half as fast)

**Comment by RalfJung** [comment]
> I would suggest that any changes that focus on potential runtime performance improvements due to assumes are a part of a separate PR.

So, can we land this even though it regresses some use-cases, given that it fixes some basic compiler invariants?
The lang team gave their okay on the behavior changes, and https://github.com/rust-lang/rust/pull/97652 landed.

**Comment by RalfJung** [comment]
So is this blocked on https://github.com/rust-lang/rust/pull/98332? I understood @nagisa as saying that (temporarily) regressing performance for this bugfix is okay -- not sure what T-compiler more broadly thinks though.

**Comment by oli-obk** [comment]
Once `assume` is an intrinsic, we can use

```rust

                    let discr_layout = this.tcx.layout_of(this.param_env.and(discr_ty)).unwrap();
                    if let Abi::Scalar(scalar) = discr_layout.layout.abi() {
                        if let Primitive::Int(..) = scalar.primitive() {
                            let range = scalar.valid_range(&this.tcx);
                            if !scalar.is_always_valid(&this.tcx)
                                && range.end >= range.start
                            {

                                let mut assume = |op, val| {
                                    let val = ConstantKind::Val(val, discr_ty);
                                    let val = Operand::Constant(Box::new(Constant {
                                        literal: val,
                                        span: source.span,
                                        user_ty: None,
                                    }));
                                    let cond = Rvalue::BinaryOp(op, Box::new((Operand::Copy(discr), val)));
                                    let cond_place = this.temp(this.tcx.types.bool, source.span);

                                    this.cfg.push_assign(
                                        block,
                                        source_info,
                                        cond_place,
                                        cond,
                                    );

                                    let func = Operand::function_handle(
                                        this.tcx,
                                        this.tcx.require_lang_item(LangItem::Assume, Some(source.span)),
                                        ty::List::empty(),
                                        source.span,
                                    );
                                    let destination = this.temp(this.tcx.types.unit, source.span);
                                    let args = vec![Operand::Move(cond_place)];

                                    let success = this.cfg.start_new_block();

                                    this.cfg.terminate(
                                        block,
                                        source_info,
                                        TerminatorKind::Call {
                                            func,
                                            args,
                                            cleanup: None,
                                            destination,
                                            target: Some(success),
                                            from_hir_call: false,
                                            fn_span: source.span,
                                        },
                                    );
                                    this.diverge_from(block);
                                    block = success;
                                    unpack!(success.unit())
                                };

                                // We want `table[e as usize ± k]` to not
                                // have bound checks, and this is the most
                                // convenient place to put the `assume`s.
                                if range.start > 0 {

                                    let enum_value_lower_bound = ConstValue::Scalar(Scalar::from_uint(range.start, discr_layout.size));

                                    assume(BinOp::Ge, enum_value_lower_bound);
                                }

                                let enum_value_upper_bound = ConstValue::Scalar(Scalar::from_uint(range.end, discr_layout.size));
                                assume(BinOp::Le, enum_value_upper_bound);
                            }
                        }
                    }
```

in `as_rvalue.rs` to get the optimization back.

File: compiler/rustc_codegen_ssa/src/mir/rvalue.rs
Comment (RalfJung) [inline compiler/rustc_codegen_ssa/src/mir/rvalue.rs:289]: Would it make sense to have a `is_signed` method on `IntTy`?

**Review by RalfJung — APPROVED** [review]
LGTM (r=me), modulo the question and also maybe changing some/all of these new `instrument` to `trace`

**Pull request merged**
